File: crates/skald/skald-workflow/examples/from_builder.rs
//! Build a workflow with the Rust API and run it.

use std::collections::HashMap;

use skald_agent::Agent;
use skald_workflow::{Workflow, WorkflowInput};

mod common;
use common::{mock_registry, plan_prompt, write_prompt};

fn main() -> anyhow::Result<()> {
    let planner = Agent::new(plan_prompt()).name("planner");
    let writer = Agent::new(write_prompt()).name("writer");
    let wf = Workflow::sequential("research", vec![planner, writer])?.with_version("0.1.0");
    wf.save("/tmp/research.yaml")?;

    let providers = mock_registry(&[
        r#"{"summary":"mock summary","steps":["read","write"]}"#,
        "Final local brief",
    ]);
    let run = wyrd_runtime::runtime().block_on(wf.run_with(
        &providers,
        WorkflowInput::from(HashMap::from([(
            "topic".to_owned(),
            "the Rust borrow checker".to_owned(),
        )])),
    ))?;
    println!("steps completed: {}", run.tasks.len());
    println!("parameters: {:?}", run.parameters);
    if let Some(text) = run.final_output.as_deref() {
        println!("final: {text}");
    }
    Ok(())
}

File: sdks/wyrd-sdk-python/examples/from_builder.py
"""Build a workflow with the Python API and execute it locally."""

from wyrd import Agent, Prompt, Workflow, tool


@tool
def web_search(query: str) -> str:
    """Search the web."""
    return f"results for {query}"


def main() -> None:
    planner = Agent(
        name="planner",
        prompt=Prompt(
            provider="mock",
            model="mock-model",
            messages=[r'{"summary":"mock summary","steps":["search","write"]}'],
            output={"summary": str, "steps": list[str]},
        ),
        tools=[web_search],
    )
    writer = Agent(
        name="writer",
        prompt=Prompt(
            provider="mock",
            model="mock-model",
            messages=["Draft a brief on: ${summary}"],
        ),
    )
    wf = Workflow.sequential("research", planner, writer)
    run = wf.run({"topic": "climate change"})
    print(f"completed {len(run.outcomes)} steps")
    print(f"cross-step params: {run.parameters}")
    print(f"final output: {run.final_output}")


if __name__ == "__main__":
    main()
